custom_components/tesla_custom/binary_sensor.py

```diff
@@ -1,40 +1,187 @@
-            )
-            for device in hass.data[TESLA_DOMAIN][config_entry.entry_id]["devices"][
-                "binary_sensor"
-            ]
-        ],
-        True,
-    )
+_LOGGER = logging.getLogger(__name__)
 
 
-class TeslaBinarySensor(TeslaDevice, BinarySensorEntity):
-    """Implement an Tesla binary sensor for parking and charger."""
+async def async_setup_entry(hass: HomeAssistant, config_entry, async_add_entities):
+    """Set up the Tesla selects by config_entry."""
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]["coordinator"]
+    cars = hass.data[DOMAIN][config_entry.entry_id]["cars"]
+    energysites = hass.data[DOMAIN][config_entry.entry_id]["energysites"]
+    entities = []
+
+    for car in cars.values():
+        entities.append(TeslaCarParkingBrake(hass, car, coordinator))
+        entities.append(TeslaCarOnline(hass, car, coordinator))
+        entities.append(TeslaCarChargerConnection(hass, car, coordinator))
+        entities.append(TeslaCarCharging(hass, car, coordinator))
+
+    for energysite in energysites.values():
+        if energysite.resource_type == RESOURCE_TYPE_BATTERY:
+            entities.append(TeslaEnergyBatteryCharging(hass, energysite, coordinator))
+            entities.append(TeslaEnergyGridStatus(hass, energysite, coordinator))
+
+    async_add_entities(entities, True)
+
+
+class TeslaCarParkingBrake(TeslaCarEntity, BinarySensorEntity):
+    """Representation of a Tesla car parking brake binary sensor."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        car: TeslaCar,
+        coordinator: TeslaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize parking brake entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "parking brake"
+        self._attr_icon = "mdi:car-brake-parking"
+        self._attr_device_class = None
 
     @property
-    def device_class(self):
-        """Return the class of this binary sensor."""
-        return (
-            self.tesla_device.sensor_type
-            if self.tesla_device.sensor_type in DEVICE_CLASSES
-            else None
-        )
+    def is_on(self):
+        """Return True if car shift state in park."""
+        return self._car.shift_state == "P"
+
+
+class TeslaCarChargerConnection(TeslaCarEntity, BinarySensorEntity):
+    """Representation of a Tesla car charger connection binary sensor."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        car: TeslaCar,
+        coordinator: TeslaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize charger connection entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "charger"
+        self._attr_icon = "mdi:ev-station"
+        self._attr_device_class = None
+
+    @property
+    def is_on(self):
+        """Return True if charger connected."""
+        return self._car.charging_state != "Disconnected"
```

Comment:
I’m seeing this sensor showing it’s “On” when I’m not connected to a charger. There must be some other states that the Tesla API is reporting besides just “Disconnected” when a charger is not connected. Anyone happen to know all the possible states?

Reply:
Here are some other states:
https://github.com/timdorr/tesla-api/issues/207

Disconnected
NoPower
Charging
Complete
Stopped

Reply:
Perfect, thanks! My only guess is "NoPower" is probably being set at times which is causing this binary sensor to show as "On" when there's not actually a charger connected. I'm still seeing this behavior so I might just return `True` when the state is either "Charging", "Complete" or "Stopped", so anything else would show up as "Off" (disconnected). The issue is I have a feeling "NoPower" might also be set when there is a charger connected, like if the car goes to sleep? Not sure...

Reply:
The intent of the sensor was if the cable was physically connected. If we change the behavior, it will need to be explained as a breaking change.

Reply:
Yeah the issue is it's not currently doing that all the time. It's showing that a cable is connected (i.e. "On") when my car is not physically connected to a charger cable. My only guess as to what's happening is something other than "Disconnected" is being reported when a charger cable is not connected, like "NoPower" (only one that would make sense). The only problem is I have a feeling "NoPower" might be also reported when there is a charger cable connected as well.

Here's my car from today. Between the first red section and the last red section, it was sitting in a parking lot the entire day so it should be reporting red ("Off") the entire time. 

<img width="510" alt="Screen Shot 2022-09-09 at 7 14 57 PM" src="https://user-images.githubusercontent.com/32663154/189465030-eb03cd51-2493-448d-9b79-4c0c728e8a65.png">

Reply:
I'd say leave the debug logs running while you're doing this and then double check the state from teslajsonpy at the time you're seeing the problem. You won't know till you see the data.

Reply:
It appears when the car goes to sleep, `None` is being reported which is causing the issue. This seems to occur with several other properties as well. Perhaps if we save the previous state, we can check if the new state is `None` and if so, just return the previous state.